adam/learner/object_recognizer.py

```diff
@@ -725,7 +754,142 @@ def replace_match_with_object_graph_node(
+                        if (
+                            perception_digraph.degree(matched_subgraph_node_successor)
+                            != 1
+                        ):
+                            raise_graph_exception(
+                                f"Node {matched_subgraph_node_successor} "
+                                f"appears to be a duplicate property node, "
+                                f"but has degree != 1",
+                                current_perception,
+                            )
+                        duplicate_nodes_to_remove.append(matched_subgraph_node_successor)
+                        continue
+                    else:
+                        external_properties.add(matched_subgraph_node_successor[0])
+
+                perception_digraph.add_edge(
+                    matched_object_node, matched_subgraph_node_successor, label=edge_label
+                )
+
+        # If there is an edge to the matched sub-graph from a node outside it,
+        # also add an edge to the object match node from that node.
+        for matched_subgraph_node_predecessor in perception_digraph.predecessors(
+            matched_subgraph_node
+        ):
+            edge_label = _get_edge_label(
+                perception_digraph,
+                matched_subgraph_node_predecessor,
+                matched_subgraph_node,
+            )
+
+            # don't want to add edges which are internal to the matched sub-graph
+            if matched_subgraph_node_predecessor not in matched_subgraph_nodes:
+                if edge_equals_ignoring_temporal_scope(edge_label, HAS_PROPERTY_LABEL):
+                    # Prevent multiple `has-property` assertions to the same color node
+                    # On a recognized object
+                    if matched_subgraph_node_predecessor[0] in external_properties:
+                        if (
+                            perception_digraph.degree(matched_subgraph_node_predecessor)
+                            != 1
+                        ):
+                            raise_graph_exception(
+                                f"Node {matched_subgraph_node_predecessor} "
+                                f"appears to be a duplicate property node, "
+                                f"but has degree != 1",
+                                current_perception,
+                            )
+                        duplicate_nodes_to_remove.append(
+                            matched_subgraph_node_predecessor
+                        )
+                        continue
+                    else:
+                        external_properties.add(matched_subgraph_node_predecessor[0])
+
+                perception_digraph.add_edge(
+                    matched_subgraph_node_predecessor,
+                    matched_object_node,
+                    label=edge_label,
+                )
+
+    if root not in SHARED_WORLD_ITEMS:
+        perception_digraph.remove_node(root)
```

Comment:
Is this the behavior we want? I assumed so based on the original function's behavior.

Reply:
Yes, that makes sense. I made the change.